Added generic math support.

MineMaarten opened the pull request:
Resolves https://github.com/nodatime/nodatime/issues/1693.

Choices made / possible discussion points:
- Ignored `internal` types.
- Added only a test for the `Period` unary negation operator, as that one actually has some implementation.
- Did not implement `IMultiplicativeIdentity`
  - `Duration` and `Period` are the most likely candidates to implement it as suggested [here](https://github.com/nodatime/nodatime/issues/1693#issue-1299781520), however I don't think there is a logical multiplicative identity for either of those (1 tick / 1 second / 1 hour / ....).
- Did not implement `IMinMaxValue` for `Period` as I could not figure out a logical `MaxValue`.
- Did not implement `IMinMaxValue` for `LocalDate` and `LocalDateTime`, as they both have `MinIsoValue` and `MaxIsoValue` already and they suggest requiring explicit usage of Iso. Let me know if implemention is desired anyway.
- Did not implement `IAdditiveIdentity` for `Instant`, as you cannot sum two Instants. However, following the rules of 'Additive Identity', I think using `new Instant(0, 0)` is technically correct.
- Implemented `IMinMaxValue` for `Offset`, however, as the interface requires properties, and the `Offset` types already has fields with the same name, I am unsure about the compatibility of this. I can remove implementing `IMinMaxValue` if desired.

jskeet commented:
Just to set expectations, I'm not sure when I'll get to this - I've got a few busy weeks ahead. It's much appreciated though, and I look forward to getting to it eventually...

jskeet commented:
Sorry about taking so long to get to this. This all looks fine. Are you happy to rebase to HEAD, and then I'll merge, or would you like me to?

At some point I might add some tests just for the sake of showing the utility of this, but that can definitely come later.

jskeet commented:
@MineMaarten: Ping for this?

MineMaarten commented:
Sorry for the late reaction. Rebased now, and added IMinMaxValue for Period as the MinValue and MaxValue properties were added with another PR in the meantime.

jskeet reviewed (approved):
This is great, thanks. I love how little actual implementation code you've had to add :)

MineMaarten commented:
All due to you adhering to .NET's naming principles like `MinValue` as well as usage of operator overloading :), as they fit the generic math operators perfectly :).

Thanks for merging!
